adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
+        """
+        return self._class_index_attribute
 
-        self._index = pd.MultiIndex.from_arrays(
-            [class_index, array_index], names=["class_index", "array_index"]
+    @class_index_attribute.setter
+    def class_index_attribute(self, value):
+        raise ValueError(
+            "Cannot set class_index_attribute directly. Use set_index() instead."
         )
 
-        return
+    @class_index_attribute.deleter
+    def class_index_attribute(self):
+        raise ValueError(
+            "Cannot delete class_index_attribute. Use set_index() with no arguments instead."
+        )
 
-    def __getitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
-        copy = deepcopy(self)
+    @property
+    def member_index(self):
+        """
+        Index of the class members.
+        """
+        return self._member_index
 
-        for k, v in self.__dict__.items():
-            if isinstance(
-                v,
-                (np.ndarray, np.ma.masked_array, list, Time, Indexable, pd.MultiIndex),
-            ):
-                copy.__dict__[k] = v[ind]
-            elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
-                copy.__dict__[k] = v
-            elif v is None:
-                pass
-            else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+    @member_index.setter
+    def member_index(self, value):
+        raise ValueError("Cannot set member_index directly. Use set_index() instead.")
 
-        return copy
+    @member_index.deleter
+    def member_index(self):
+        raise ValueError(
+            "Cannot delete member_index. Use set_index() with no arguments instead."
+        )
 
-    def __delitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
+    @property
+    def member_length(self):
+        """
+        Length of the class members.
+        """
+        return self._member_length
+
+    @member_length.setter
+    def member_length(self, value):
+        raise ValueError("Cannot set member_length directly. Use set_index() instead.")
 
+    @member_length.deleter
+    def member_length(self):
+        raise ValueError(
+            "Cannot delete member_length. Use set_index() with no arguments instead."
+        )
+
+    def set_index(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        """
+        Set an index on the class for the given values or attribute name.
+
+        Parameters
+        ----------
+        values : str, np.ArrayLike
+            The values to be indexed. If a string is given then the values are taken from the
+            attribute of the same name. If an array is given then the values are taken from the
+            array itself. If None is given then the index is set to the range of length
+            of the class's sliceable members.
```

Comment:
This level of type-based overloading seems pretty extreme to me. Why are all three behaviors under one method, as opposed to
```
def index_by(self, attrib_name: str):
    ...
  
def reindex(self, ordering: npt.ArrayLike):
    ...

# give this a better name - I don't understand what it does.
def index_by_none(self):
    ...